core/finality-grandpa/src/lib.rs

```diff
@@ -189,6 +191,58 @@ type CommunicationOutH<Block, H> = grandpa::voter::CommunicationOut<
 	AuthorityId,
 >;
 
+/// Aura key provider for offchain workers.
+pub struct GrandpaKeyProvider<Block, Client> {
+	_marker: PhantomData<Block>,
+	client: Arc<Client>,
+	keystore: Arc<LocalStore<fg_primitives::AuthorityPair>>,
+}
+
+impl<Block, Client> GrandpaKeyProvider<Block, Client>
+where
+	Block: BlockT,
+	Client: ProvideRuntimeApi + Send + Sync,
+	Client::Api: GrandpaApi<Block>,
+{
+	/// Creates a new `GrandpaKeyProvider`.
+	pub fn new(
+		client: Arc<Client>,
+		keystore: Arc<LocalStore<fg_primitives::AuthorityPair>>,
+	) -> Self {
+		Self {
+			_marker: PhantomData,
+			client,
+			keystore,
+		}
+	}
+}
+
+impl<Block, Client> offchain::AuthorityKeyProvider<Block> for GrandpaKeyProvider<Block, Client>
+where
+	Block: BlockT,
+	Client: ProvideRuntimeApi + Send + Sync,
+	Client::Api: GrandpaApi<Block>,
+{
+	fn authority_key(
+		&self,
+		at: &BlockId<Block>,
+	) -> Option<Box<dyn offchain::OffchainKey>> {
+		let authorities = self.client
+			.runtime_api()
+			.grandpa_authorities(at)
```

Comment:
This won't give us the active authority set because GRANDPA authority set changes are only enacted on finality. What we need to do is figure out what the highest finalized ancestor of `at` is (or `at` itself), and query the runtime state at that block.